kani-driver/src/exe_trace.rs

```diff
@@ -0,0 +1,264 @@
+            let (exe_trace, exe_trace_func_name) =
+                format_unit_test(&harness.mangled_name, &det_vals[..], &interp_det_vals[..]);
+
+            println!("Executable trace:\n");
+            println!("{}", exe_trace);
+
+            if self.args.add_exe_trace_to_src {
+                if !self.args.quiet {
+                    println!("Now modifying the source code to include the executable trace.");
+                }
+                let proof_harness_end_line: usize = harness
+                    .original_end_line
+                    .parse()
+                    .expect("Couldn't convert proof harness line from str to usize");
+                self.modify_src_code(
+                    &harness.original_file,
+                    proof_harness_end_line,
+                    &exe_trace,
+                    &exe_trace_func_name,
+                );
+            } else {
+                println!(
+                    "To automatically add this executable trace to the src code, run Kani with `--add-exe-trace-to-src`."
+                );
+            }
+        }
+    }
+
+    /// Add the exe trace to the user's source code, format it, and short circuit if code already present.
+    fn modify_src_code(
+        &self,
+        src_file_path: &str,
+        proof_harness_end_line: usize,
+        exe_trace: &str,
+        exe_trace_func_name: &str,
+    ) {
+        // Prep the source code and exec trace func.
+        let src_as_str =
+            fs::read_to_string(src_file_path).expect("Couldn't access proof harness source file");
+        let mut src_lines = src_as_str.split('\n').collect::<Vec<&str>>();
+        let mut exe_trace_lines = exe_trace.split('\n').collect::<Vec<&str>>();
+
+        // Short circuit if exe trace already in source code.
+        if src_as_str.contains(exe_trace_func_name) {
+            if !self.args.quiet {
+                println!("Exe trace already found in source code, so skipping modification.");
+            }
+            return;
+        }
+
+        // Calc inserted indexes and line numbers into source code to rustfmt only those lines.
+        // Indexes are into the vector (0-idx), lines are into src file (1-idx).
+        let start_line = proof_harness_end_line + 1;
+        let start_idx = start_line - 1;
+        // If start_line=2, exe_trace.len()=3, then inserted code ends at line 4 (start + len - 1).
+        let end_line = start_line + exe_trace_lines.len() - 1;
+
+        // Splice the exec trace func into the proof harness source code.
+        // start_idx is at max len(src_lines), so no panic here, even if proof harness is at the end of src file.
+        let mut src_right = src_lines.split_off(start_idx);
+        src_lines.append(&mut exe_trace_lines);
+        src_lines.append(&mut src_right);
+        let new_src = src_lines.join("\n");
+        fs::write(src_file_path, new_src).expect("Couldn't write to proof harness source file");
```

Comment:
You should probably write to a new file, run rustfmt and move the new file to the src_file_path location at the end. Otherwise, if any of this process fails, this could corrupt the user source code.

At least on Linux, moving a file should be atomic.

Reply:
I modified the code to create a temporary file and rename that file to the actual source file.